File: adapters/jvm/stdlib/src/main/kotlin/mud/stdlib/world/Room.kt
package mud.stdlib.world

open class Room : GameObject() {
    private val exits = mutableMapOf<String, String>()

    fun exit(direction: String, to: String) {
        exits[direction] = to
    }

    fun exits(): Map<String, String> = exits
    fun hasExit(direction: String): Boolean = direction in exits

    open fun onEnter(player: String) {}
}

File: bootstrap/ruby/stdlib/world/room.rb
# frozen_string_literal: true

require_relative 'game_object'

module MudAdapter
  module Stdlib
    module World
      class Room < GameObject
        class << self
          def exit(direction, to:)
            @exits ||= {}
            @exits[direction] = to
          end

          def exits
            @exits ||= {}
          end

          def inherited(subclass)
            super
            subclass.instance_variable_set(:@exits, @exits&.dup || {})
          end
        end

        def exits
          self.class.exits
        end

        def exit_directions
          exits.keys
        end

        def has_exit?(direction)
          exits.key?(direction)
        end

        def on_enter(player)
          # default no-op, subclasses override
        end
      end
    end
  end
end
